src/backend/distributed/shared_library_init.c

```diff
@@ -2146,12 +2167,69 @@ StatisticsCollectionGucCheckHook(bool *newval, void **extra, GucSource source)
 static void
 CitusAuthHook(Port *port, int status)
 {
+	if (!IsCitusInternalBackend())
+	{
+		/*
+		 * We raise the shared connection counter pre-emptively. As a result, we may
+		 * have scenarios in which a few simultaneous connection attempts prevent
+		 * each other from succeeding, but we avoid scenarios where we oversubscribe
+		 * the system.
+		 *
+		 * By also calling RegisterClientBackendCounterDecrement here, we immediately
+		 * lower the counter if we throw a FATAL error below. The client connection
+		 * counter may temporarily exceed maxClientConnections in between.
+		 */
+		RegisterClientBackendCounterDecrement();
+		IncrementClientBackendCounter();
+
+		/*
+		 * Limit non-superuser client connections if citus.max_client_connections
+		 * is set.
+		 */
+		if (MaxClientConnections >= 0 &&
```

Comment:
I guess our behavior is slightly different compared to Postgres' `max_connections`, which also includes the `superuser_reserved_connections`.

In our case, superusers are even more flexible, can exceed `max_client_connections`.

I think that is fine, just wanted to bring up if you don't agree